packages/ee/billing/backend/usage.service.ts

```diff
@@ -0,0 +1,77 @@
+import { billingService } from "./billing.service";
+import { ActivepiecesError, ErrorCode, FlowVersion, ProjectId, Trigger, Action, apId } from "@activepieces/shared";
+import { databaseConnection } from "@backend/database/database-connection";
+import { createRedisLock } from "@backend/database/redis-connection";
+import { ProjectUsage } from "../shared/usage";
+import { ProjectUsageEntity } from "./usage.entity";
+import { logger } from "@backend/helper/logger";
+import dayjs from "dayjs";
+
+const projectUsageRepo = databaseConnection.getRepository<ProjectUsage>(ProjectUsageEntity);
+
+export const usageService = {
+    async limit(request: { projectId: ProjectId; flowVersion: FlowVersion; }): Promise<{ perform: true }> {
+        const quotaLock = await createRedisLock(5 * 1000);
+        try {
+            quotaLock.acquire(`usage_${request.projectId}}`);
+            const projectUsage = await usageService.getUsage({ projectId: request.projectId });
+            const numberOfSteps = countSteps(request.flowVersion);
+            const projectPlan = await billingService.getPlan({ projectId: request.projectId });
+            if (projectUsage.consumedTasks + numberOfSteps > projectPlan.tasks) {
+                throw new ActivepiecesError({
+                    code: ErrorCode.TASK_QUOTA_EXCEEDED,
+                    params: { id: request.projectId },
+                });
+            }
+            projectUsage.consumedTasks += numberOfSteps;
+            await projectUsageRepo.save(projectUsage);
+        } catch (e) {
+            // TODO add sentry, Don't block for internal error, just continue and log it for now.
+            logger.error(e);
+        } finally {
+            quotaLock.release();
+        };
+        return {
+            perform: true,
+        }
+    },
+    async getUsage({ projectId }: { projectId: ProjectId }): Promise<ProjectUsage> {
+        let projectUsage = await projectUsageRepo.findOneBy({ projectId });
+        if (projectUsage === undefined || projectUsage === null || isPastResetDate(projectUsage.nextResetDatetime)) {
+            const plan = await billingService.getPlan({ projectId });
+            projectUsage = await projectUsageRepo.save({
+                id: apId(),
+                projectId,
+                consumedTasks: 0,
+                nextResetDatetime: nextResetDatetime(plan.subscriptionStartDatetime),
+            });
+        }
+        return projectUsage;
+    },
+}
+
+
+function countSteps(flowVersion: FlowVersion): number {
+    let steps = 0;
+    let currentStep: Trigger | Action | undefined = flowVersion.trigger;
+    while (currentStep !== undefined) {
+        currentStep = currentStep.nextAction;
+        steps++;
+    }
+    return steps;
+}
```

Comment:
Shouldn't step count be calculated one time on flow creation?

Reply:
Yes, I will follow up on this on branching branch.